rust/parquet/src/arrow/converter.rs

```diff
@@ -0,0 +1,219 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::arrow::record_reader::RecordReader;
+use crate::data_type::DataType;
+use arrow::array::Array;
+use arrow::compute::cast;
+use std::convert::From;
+use std::sync::Arc;
+
+use crate::errors::Result;
+use arrow::datatypes::ArrowPrimitiveType;
+
+use arrow::array::ArrayDataBuilder;
+use arrow::array::PrimitiveArray;
+use std::marker::PhantomData;
+
+use crate::data_type::BoolType;
+use crate::data_type::DoubleType as ParquetDoubleType;
+use crate::data_type::FloatType as ParquetFloatType;
+use crate::data_type::Int32Type as ParquetInt32Type;
+use crate::data_type::Int64Type as ParquetInt64Type;
+use arrow::datatypes::BooleanType;
+use arrow::datatypes::Float32Type;
+use arrow::datatypes::Float64Type;
+use arrow::datatypes::Int16Type;
+use arrow::datatypes::Int32Type;
+use arrow::datatypes::Int64Type;
+use arrow::datatypes::Int8Type;
+use arrow::datatypes::UInt16Type;
+use arrow::datatypes::UInt32Type;
+use arrow::datatypes::UInt64Type;
+use arrow::datatypes::UInt8Type;
+
+/// A converter is used to consume record reader's content and convert it to arrow
+/// primitive array.
+pub trait Converter<T: DataType> {
+    /// This method converts record reader's buffered content into arrow array.
+    fn convert(record_reader: &mut RecordReader<T>) -> Result<Arc<Array>>;
+}
+
+/// Cast converter first converts record reader's buffer to arrow's
+/// `PrimitiveArray<ArrowSourceType>`, then casts it to `PrimitiveArray<ArrowTargetType>`.
+pub struct CastConverter<ParquetType, ArrowSourceType, ArrowTargetType> {
+    _parquet_marker: PhantomData<ParquetType>,
+    _arrow_source_marker: PhantomData<ArrowSourceType>,
+    _arrow_target_marker: PhantomData<ArrowTargetType>,
+}
+
+impl<ParquetType, ArrowSourceType, ArrowTargetType> Converter<ParquetType>
+    for CastConverter<ParquetType, ArrowSourceType, ArrowTargetType>
+where
+    ParquetType: DataType,
+    ArrowSourceType: ArrowPrimitiveType,
+    ArrowTargetType: ArrowPrimitiveType,
+{
+    fn convert(record_reader: &mut RecordReader<ParquetType>) -> Result<Arc<Array>> {
```

Comment:
Was looking at how we can avoid iterating this value by value - seems in C++ it just does a `reinterpret_cast` on the values decoded from Parquet, which is interesting: https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/reader_internal.cc#L575

Reply:
This is used when arrow's data type is same as parquet's physical type. In cast kernel's implements, the one by one conversion only happens when ArrowSourceType is different from ArrowTargetType.

Reply:
No. This is also used when Arrow type is narrower than Parquet type, for instance, converting Parquet's Int32 to Arrow's UInt8. See https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/reader_internal.cc#L1043.

I think what you have is a good start - we can think about ways to optimize this later.

Reply:
I think the magic hides in std::copy